### PR TITLE
WIP relove empty value "" in headers

### PR DESCRIPTION
**Pull request by MouceL**
**Please provide a description of this PR:**

resolve the problems describe here https://github.com/istio/istio/issues/35489

### CODE REVIEW

**Comment by MouceL** [comment]
@googlebot I signed it!

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 1]
```diff
@@ -988,17 +988,23 @@ func translateHeaderMatch(name string, in *networking.StringMatch) *route.Header
 
 	switch m := in.MatchType.(type) {
 	case *networking.StringMatch_Exact:
-		out.HeaderMatchSpecifier = &route.HeaderMatcher_ExactMatch{ExactMatch: m.Exact}
+		if m.Exact != "" {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:991]: this doesn't seem to solve it, just makes it not do anything.

https://github.com/istio/istio/blob/90e61d212dce2550e29250afb053ed45099adb6a/pilot/pkg/networking/core/v1alpha3/route/route.go#L923 is the code intended to solve this I believe. I think its incorrectly checking `nil` when `""` can occur as well?

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:991]: Although I think maybe this should be invalid and rejected in the webhook.

Comment (MouceL) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:991]: > this doesn't seem to solve it, just makes it not do anything.
> 
> https://github.com/istio/istio/blob/90e61d212dce2550e29250afb053ed45099adb6a/pilot/pkg/networking/core/v1alpha3/route/route.go#L923
> 
> is the code intended to solve this I believe. I think its incorrectly checking `nil` when `""` can occur as well?

  In the absence of any header match specifier, match will default to present_match, so it seems work in envoy when you set  prefix: ""  
headers:
    x-my-header:
      prefix: "" # Empty value means only the presence of the header is checked

==>
          "match": {
           "path": "/productpage",
           "case_sensitive": true,
           "headers": [
            {
             "name": "x-my-header"
            }
           ]
          }

Comment (MouceL) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:991]: according `(repeated config.route.v3.HeaderMatcher) Specifies a set of headers that the route should match on. The router will check the request’s headers against all the specified headers in the route config. A match will happen if all the headers in the route are present in the request with the same values (or based on presence if the value field is not in the config)`  ==>  so  if headers only contain name, the match will default to present_match , it's envoy feature

Comment (MouceL) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:991]: @cywang1905 hello

Comment (cywang1905) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:991]: = =

**Comment by cywang1905** [comment]
@MouceL = =

**Pull request closed without merging**